Pubsub: handle stream closed

After reviewing the PR https://github.com/libp2p/py-libp2p/pull/299#discussion_r326230742,   I found that the stream exceptions are not handled completely in the current codebase. We should revisit it and add tests for the situation when a pubsub stream is closed/reset.

### Reference
- Go implementation: https://github.com/libp2p/go-libp2p-pubsub/blob/master/comm.go#L38-L46

Example exception I'm getting while trying to write a message:
```

2019-10-14 10:17:57 [ERROR] P2P.peers: Can't publish alive message
Traceback (most recent call last):
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/network/stream/net_stream.py", line 59, in write
    return await self.muxed_stream.write(data)
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/stream_muxer/mplex/mplex_stream.py", line 149, in write
    raise MplexStreamClosed(f"cannot write to closed stream: data={data}")
libp2p.stream_muxer.mplex.exceptions.MplexStreamClosed: cannot write to closed stream: data=b'\xb0\x01\x12\xad\x01\n"\x12 \xa7J\xb1\x
f0\xcaj\x84@\xe1\xa27\xe2\x0bv\x15t;\xe2\x13!\x0e\x9c\x8d\x13{\xb0[\x06\x1bA\xfcY\x12v{"address":"http://***","intere
sts":null,"peer_type":"HTTP","version":"0.0.post0.dev5+g8f7c4e8.dirty"}\x1a\x08\x15\xcd\x08\xa3d\xe9\x0e1"\x05ALIVE'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jon/virtualenvs/pyaleph/pyaleph/src/aleph/services/p2p/peers.py", line 27, in publish_host
    await psub.publish(topic, msg)
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/pubsub/pubsub.py", line 409, in publish
    await self.push_msg(self.host.get_id(), msg)
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/pubsub/pubsub.py", line 481, in push_msg
    await self.router.publish(msg_forwarder, msg)
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/pubsub/floodsub.py", line 88, in publish
    await stream.write(encode_varint_prefixed(rpc_msg.SerializeToString()))
  File "/home/jon/virtualenvs/pyaleph/py-libp2p/libp2p/network/stream/net_stream.py", line 61, in write
    raise StreamClosed from error
libp2p.network.stream.exceptions.StreamClosed
```

As a side note, this means that my routines that periodically try to reconnect to host to which we have been disconnected from, fails (as a call to host.connect(multinfo) returns immediately if the host is still in the connected list).
We should perhaps disconnect from it on the host/swarm side when this happens?

--- a/libp2p/network/connection/swarm_connection.py
+++ b/libp2p/network/connection/swarm_connection.py
@@ -43,6 +43,9 @@ class SwarmConn(INetConn):
         # We *could* optimize this but it really isn't worth it.
         for stream in self.streams:
             await stream.reset()
+        # Force context switch for stream handlers to process the stream reset event we just emit
+        # before we cancel the stream handler tasks.
+        await asyncio.sleep(0.1)
 
         for task in self._tasks:
             task.cancel()

--- a/libp2p/network/swarm.py
+++ b/libp2p/network/swarm.py
@@ -248,8 +248,8 @@ class Swarm(INetwork):
         # TODO: Should be changed to close multisple connections,
         #   if we have several connections per peer in the future.
         connection = self.connections[peer_id]
-        # NOTE: `connection.close` performs `del self.connections[peer_id]` for us,
-        #   so we don't need to remove the entry here.
+        # NOTE: `connection.close` will perform `del self.connections[peer_id]`
+        # and `notify_disconnected` for us.
         await connection.close()
 
         logger.debug("successfully close the connection to peer %s", peer_id)

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -60,6 +60,7 @@ class Pubsub:
     router: "IPubsubRouter"
 
     peer_queue: "asyncio.Queue[ID]"
+    dead_peer_queue: "asyncio.Queue[ID]"
 
     protocols: List[TProtocol]
 
@@ -100,7 +101,10 @@ class Pubsub:
 
         # Register a notifee
         self.peer_queue = asyncio.Queue()
-        self.host.get_network().register_notifee(PubsubNotifee(self.peer_queue))
+        self.dead_peer_queue = asyncio.Queue()
+        self.host.get_network().register_notifee(
+            PubsubNotifee(self.peer_queue, self.dead_peer_queue)
+        )
 
         # Register stream handlers for each pubsub router protocol to handle
         # the pubsub streams opened on those protocols

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -253,10 +253,10 @@ class Pubsub:
         try:
             await self.continuously_read_stream(stream)
         except StreamEOF as error:
-            stream.close()
+            await stream.close()
             logger.debug("fail to read from peer %s, error=%s", peer_id, error)
         except (ParseError, IncompleteReadError, StreamReset) as error:
-            stream.reset()
+            await stream.reset()
             logger.debug("read corrupted data from peer %s, error=%s", peer_id, error)
 
     async def _handle_new_peer(self, peer_id: ID) -> None:
@@ -281,7 +281,7 @@ class Pubsub:
 
         logger.debug("added new peer %s", peer_id)
 
-    async def _handle_dead_peer(self, peer_id: ID) -> None:
+    def _handle_dead_peer(self, peer_id: ID) -> None:
         del self.peers[peer_id]
 
         for topic in self.peer_topics:
@@ -300,14 +300,9 @@ class Pubsub:
         pubsub protocols we support
         """
         while True:
-
             peer_id: ID = await self.peer_queue.get()
-
             # Add Peer
-
             asyncio.ensure_future(self._handle_new_peer(peer_id))
-            # Force context switch
-            await asyncio.sleep(0)
 
     async def handle_dead_peer_queue(self) -> None:
         """
@@ -315,14 +310,9 @@ class Pubsub:
         remove peer info from pubsub and pubsub router.
         """
         while True:
-
             peer_id: ID = await self.dead_peer_queue.get()
-
             # Remove Peer
-            asyncio.ensure_future(self._handle_dead_peer(peer_id))
-
-            # Force context switch
-            await asyncio.sleep(0)
+            self._handle_dead_peer(peer_id)
 
     def handle_subscription(
         self, origin_id: ID, sub_message: rpc_pb2.RPC.SubOpts

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -141,6 +141,7 @@ class Pubsub:
 
         # Call handle peer to keep waiting for updates to peer queue
         asyncio.ensure_future(self.handle_peer_queue())
+        asyncio.ensure_future(self.handle_dead_peer_queue())
 
     def get_hello_packet(self) -> rpc_pb2.RPC:
         """Generate subscription message with all topics we are subscribed to

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -248,16 +248,17 @@ class Pubsub:
         """
         peer_id = stream.muxed_conn.peer_id
 
-        # Error handling pattern reference:
-        # https://github.com/libp2p/go-libp2p-pubsub/blob/534fe2f382d8dd75dab89ddb0760542546c9f24e/comm.go#L38-L46  # noqa: E501
         try:
             await self.continuously_read_stream(stream)
-        except StreamEOF as error:
-            await stream.close()
-            logger.debug("fail to read from peer %s, error=%s", peer_id, error)
-        except (ParseError, IncompleteReadError, StreamReset) as error:
+        except (StreamEOF, StreamReset, ParseError, IncompleteReadError) as error:
+            logger.debug(
+                "fail to read from peer %s, error=%s,"
+                "closing the stream and remove the peer from record",
+                peer_id,
+                error,
+            )
             await stream.reset()
-            logger.debug("read corrupted data from peer %s, error=%s", peer_id, error)
+            self._handle_dead_peer(peer_id)
 
     async def _handle_new_peer(self, peer_id: ID) -> None:
         try:
@@ -282,6 +283,8 @@ class Pubsub:
         logger.debug("added new peer %s", peer_id)
 
     def _handle_dead_peer(self, peer_id: ID) -> None:
+        if peer_id not in self.peers:
+            return
         del self.peers[peer_id]
 
         for topic in self.peer_topics:

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -79,6 +79,8 @@ class Pubsub:
     # TODO: Be sure it is increased atomically everytime.
     counter: int  # uint64
 
+    _tasks: List["asyncio.Future[Any]"]
+
     def __init__(
         self, host: IHost, router: "IPubsubRouter", my_id: ID, cache_size: int = None
     ) -> None:
@@ -139,9 +141,10 @@ class Pubsub:
 
         self.counter = time.time_ns()
 
+        self._tasks = []
         # Call handle peer to keep waiting for updates to peer queue
-        asyncio.ensure_future(self.handle_peer_queue())
-        asyncio.ensure_future(self.handle_dead_peer_queue())
+        self._tasks.append(asyncio.ensure_future(self.handle_peer_queue()))
+        self._tasks.append(asyncio.ensure_future(self.handle_dead_peer_queue()))
 
     def get_hello_packet(self) -> rpc_pb2.RPC:
         """Generate subscription message with all topics we are subscribed to
@@ -174,7 +177,7 @@ class Pubsub:
                     logger.debug(
                         "received `publish` message %s from peer %s", msg, peer_id
                     )
-                    asyncio.ensure_future(self.push_msg(msg_forwarder=peer_id, msg=msg))
+                    self._tasks.append(asyncio.ensure_future(self.push_msg(msg_forwarder=peer_id, msg=msg)))
 
             if rpc_incoming.subscriptions:
                 # deal with RPC.subscriptions
@@ -305,7 +308,7 @@ class Pubsub:
         while True:
             peer_id: ID = await self.peer_queue.get()
             # Add Peer
-            asyncio.ensure_future(self._handle_new_peer(peer_id))
+            self._tasks.append(asyncio.ensure_future(self._handle_new_peer(peer_id)))
 
     async def handle_dead_peer_queue(self) -> None:
         """
@@ -537,3 +540,11 @@ class Pubsub:
         if not self.my_topics:
             return False
         return any(topic in self.my_topics for topic in msg.topicIDs)
+
+    async def close(self) -> None:
+        for task in self._tasks:
+            task.cancel()
+            try:
+                await task
+            except asyncio.CancelledError:
+                pass

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -32,6 +32,7 @@ from .validators import signature_validator
 
 if TYPE_CHECKING:
     from .pubsub_router_interface import IPubsubRouter  # noqa: F401
+    from typing import Any  # noqa: F401
 
 
 logger = logging.getLogger("libp2p.pubsub")
@@ -177,7 +178,11 @@ class Pubsub:
                     logger.debug(
                         "received `publish` message %s from peer %s", msg, peer_id
                     )
-                    self._tasks.append(asyncio.ensure_future(self.push_msg(msg_forwarder=peer_id, msg=msg)))
+                    self._tasks.append(
+                        asyncio.ensure_future(
+                            self.push_msg(msg_forwarder=peer_id, msg=msg)
+                        )
+                    )
 
             if rpc_incoming.subscriptions:
                 # deal with RPC.subscriptions
@@ -311,10 +316,8 @@ class Pubsub:
             self._tasks.append(asyncio.ensure_future(self._handle_new_peer(peer_id)))
 
     async def handle_dead_peer_queue(self) -> None:
-        """
-        Continuously read from dead peer queue and close the stream between that peer and
-        remove peer info from pubsub and pubsub router.
-        """
+        """Continuously read from dead peer queue and close the stream between
+        that peer and remove peer info from pubsub and pubsub router."""
         while True:
             peer_id: ID = await self.dead_peer_queue.get()
             # Remove Peer

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -281,6 +281,17 @@ class Pubsub:
 
         logger.debug("added new peer %s", peer_id)
 
+    async def _handle_dead_peer(self, peer_id: ID) -> None:
+        del self.peers[peer_id]
+
+        for topic in self.peer_topics:
+            if peer_id in self.peer_topics[topic]:
+                self.peer_topics[topic].remove(peer_id)
+
+        self.router.remove_peer(peer_id)
+
+        logger.debug("removed dead peer %s", peer_id)
+
     async def handle_peer_queue(self) -> None:
         """
         Continuously read from peer queue and each time a new peer is found,
@@ -295,6 +306,21 @@ class Pubsub:
             # Add Peer
 
             asyncio.ensure_future(self._handle_new_peer(peer_id))
+            # Force context switch
+            await asyncio.sleep(0)
+
+    async def handle_dead_peer_queue(self) -> None:
+        """
+        Continuously read from dead peer queue and close the stream between that peer and
+        remove peer info from pubsub and pubsub router.
+        """
+        while True:
+
+            peer_id: ID = await self.dead_peer_queue.get()
+
+            # Remove Peer
+            asyncio.ensure_future(self._handle_dead_peer(peer_id))
+
             # Force context switch
             await asyncio.sleep(0)
 

--- a/libp2p/pubsub/pubsub.py
+++ b/libp2p/pubsub/pubsub.py
@@ -163,13 +163,7 @@ class Pubsub:
         peer_id = stream.muxed_conn.peer_id
 
         while True:
-            try:
-                incoming: bytes = await read_varint_prefixed_bytes(stream)
-            except (ParseError, IncompleteReadError) as error:
-                logger.debug(
-                    "read corrupted data from peer %s, error=%s", peer_id, error
-                )
-                continue
+            incoming: bytes = await read_varint_prefixed_bytes(stream)
             rpc_incoming: rpc_pb2.RPC = rpc_pb2.RPC()
             rpc_incoming.ParseFromString(incoming)
             if rpc_incoming.publish:
@@ -252,13 +246,18 @@ class Pubsub:
 
         :param stream: newly created stream
         """
+        peer_id = stream.muxed_conn.peer_id
+
+        # Error handling pattern reference:
+        # https://github.com/libp2p/go-libp2p-pubsub/blob/534fe2f382d8dd75dab89ddb0760542546c9f24e/comm.go#L38-L46  # noqa: E501
         try:
             await self.continuously_read_stream(stream)
-        except (StreamEOF, StreamReset) as error:
-            logger.debug("fail to read from stream, error=%s", error)
-            await stream.reset()
-            # TODO: what to do when the stream is terminated?
-            # disconnect the peer?
+        except StreamEOF as error:
+            stream.close()
+            logger.debug("fail to read from peer %s, error=%s", peer_id, error)
+        except (ParseError, IncompleteReadError, StreamReset) as error:
+            stream.reset()
+            logger.debug("read corrupted data from peer %s, error=%s", peer_id, error)
 
     async def _handle_new_peer(self, peer_id: ID) -> None:
         try:

--- a/libp2p/pubsub/pubsub_notifee.py
+++ b/libp2p/pubsub/pubsub_notifee.py
@@ -15,13 +15,21 @@ if TYPE_CHECKING:
 class PubsubNotifee(INotifee):
 
     initiator_peers_queue: "asyncio.Queue[ID]"
+    dead_peers_queue: "asyncio.Queue[ID]"
 
-    def __init__(self, initiator_peers_queue: "asyncio.Queue[ID]") -> None:
+    def __init__(
+        self,
+        initiator_peers_queue: "asyncio.Queue[ID]",
+        dead_peers_queue: "asyncio.Queue[ID]",
+    ) -> None:
         """
         :param initiator_peers_queue: queue to add new peers to so that pubsub
         can process new peers after we connect to them
+        :param dead_peers_queue: queue to add dead peers to so that pubsub
+        can process dead peers after we disconnect from each other
         """
         self.initiator_peers_queue = initiator_peers_queue
+        self.dead_peers_queue = dead_peers_queue
 
     async def opened_stream(self, network: INetwork, stream: INetStream) -> None:
         pass

--- a/libp2p/pubsub/pubsub_notifee.py
+++ b/libp2p/pubsub/pubsub_notifee.py
@@ -49,7 +49,14 @@ class PubsubNotifee(INotifee):
         await self.initiator_peers_queue.put(conn.muxed_conn.peer_id)
 
     async def disconnected(self, network: INetwork, conn: INetConn) -> None:
-        pass
+        """
+        Add peer_id to dead_peers_queue, so that pubsub and its router can
+        remove this peer_id and close the stream inbetween.
+
+        :param network: network the connection was opened on
+        :param conn: connection that was opened
+        """
+        await self.dead_peers_queue.put(conn.muxed_conn.peer_id)
 
     async def listen(self, network: INetwork, multiaddr: Multiaddr) -> None:
         pass

